# Final exam (Individual)


### <p style="text-align: right;"> &#9989; Sit Soe.</p>
### <p style="text-align: right;"> &#9989; soesit.</p>

# CMSE 202 Final (Fall 2022)

The goal of this final is to give you the opportunity to test out some of the skills that you've developed having now finished CMSE 202. In particular, you'll be committing and pushing repository changes to a GitHub repository, working with data to build a network graph, performing regression analysis, and classifying data using a machine learning classifier. You should find that you have all of the skills necessary to complete this exam having completed the second half of CMSE 202!

You are encouraged to look through the entire exam before you get started so that you can appropriately budget your time and understand the broad goals of the exam. Once you've read through it, you'll probably want to make sure you do Part 1 first to ensure that your GitHub repository is working correctly. Let your instructor know right away if you run into issues!

The exam is set up so that even if you get stuck on one part there are opportunities to get points on the other parts, so consider jumping ahead if you feel like you aren't making progress and then come back later if you have time.

If you have any questions during the exam, you may ask the instructor, the TA, or the LA privately. If you are attending in-person, simply raise your hand and one of us will come over to you. If you are attending via-Zoom, please use the ask-for-help feature or privately message the instructor, the TA, or the LA. We will make any announcements or exam clarifications/corrections to the class via the following Google Document.

[LINK]

**Important note about using online resources**: This exam is "open internet". That means that you can look up documentation, google how to accomplish certain Python tasks, etc. Being able to effectively use the internet for computational modeling and data science is a very important skill, so we want to make sure you have the opportunity to exercise that skill. **However**: The use of any person-to-person communication software is absolutely not acceptable. If you are seen accessing your email, using a chat program (e.g. Slack), or any sort of collaborative cloud storage or document software (e.g. Google Documents other than the one we will make announcements on), you will be at risk for receiving a zero on the exam.

**Keep your eyes on your screen!** This exam is designed to give *you* the opportunity to show the instructor what you can do and you should hold yourself accountable for maintaining a high level of academic integrity. If any of the instructors observe suspicious behavior, you will, again, risk receiving a zero.

------
<a id="toc"></a>

## Table of contents


0. [Part 0: Upgrade packages](#part0) (1 point)

1. [Part 1: Git](#part1) (9 points)

2. [Part 2: Data Preprocessing](#part2) (20 points)

3. [Part 3: Regression](#part3) (15 points)

4. [Part 4: Principal Component Analysis](#part4) (12 points)

5. [Part 5: Support Vector Machines](#part5) (15 points)

6. [Part 6: Regression Redux](#part6) (7 points)

7. [Part 7: Conceptual Questions](#part7) (14 points)

8. [Part 8: Conclusion](#conclusion) (3 points)

In [1]:
grades = [1, 9, 20, 15, 12, 15, 7, 14, 3]

print(f"The total grade for this final is {sum(grades)}" )

The total grade for this final is 96


---
<a id="part0"></a>
[Back to ToC](#toc)

# Part 0: Upgrade Packages

**&#9989; Question 0.1 (1 point)**: Run the cell below. Do you have the correct packages ? If not upgrade them. **You must do this in order to avoid issues in the rest of the notebook.**


In [14]:

# Standard libraries
%pylab
%matplotlib inline

import seaborn as sns
import pandas as pd

# Data cleaning Libraries
from sklearn.decomposition import PCA
from sklearn.model_selection import GridSearchCV, train_test_split
from sklearn.preprocessing import PowerTransformer
from sklearn.utils import resample
# Classifier Libraries
from sklearn.linear_model import Perceptron
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix, accuracy_score, ConfusionMatrixDisplay
from sklearn.svm import SVC, LinearSVC

import statsmodels.api as sm 

from sklearn import __version__ as sk_version
sns.set_context("talk")

print(f"Sklearn version should be 1.1.3 and I have {sk_version}")


Using matplotlib backend: agg
Populating the interactive namespace from numpy and matplotlib
Sklearn version should be 1.1.3 and I have 1.1.3


---
<a id="part1"></a>
[Back to ToC](#toc)

## Part 1: Git (9 points)

For this assignment, you're going to add it to the `cmse202-f22-turnin` repository you created in class so that you can track your progress on the assignment and preserve the final version that you turn in. In order to do this you need to

**&#9989; Question 1.1 (1 point)**: Navigate to your `cmse202-f22-turnin` **local** repository and create a new directory called `final` and copy this notebook in that new directory.

``` bash
# Put the command(s) for creating the new directory

```
mkdir final

&#9989; **Question 1.2 (3 points)** Check the status of your local `git`.

``` bash 
# Put the command you used to check the status of git

```
git status 
Copy and paste below the output of the command.

``` bash
# Paste it here
```
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   ../Project /FirstYearGPA.csv

Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        deleted:    ../CMSE202_Midterm_Sect01_STUDENT.ipynb
        deleted:    ../Project /FirstYearGPA.csv
        modified:   ../hw-02/.ipynb_checkpoints/HW-02_PandasAndObjects-STUDENT-checkpoint.ipynb
        modified:   ../hw-04/HW-04_Regression-STUDENT.ipynb
        modified:   ../midterm/CMSE202_Midterm_Sect01_STUDENT.ipynb

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        ./
        ../hw-03/.ipynb_checkpoints/
        ../hw-03/asoiaf-book1-edges.csv
        ../hw-03/store_placement.py
        ../hw-04/.ipynb_checkpoints/
        ../hw-04/ads.csv
        ../hw-04/housing.csv
        ../midterm/.ipynb_checkpoints/
        ../midterm/metadata.csv
        ../midterm/spectra_data.csv 

What is the name of the branch you are in ? 

``` bash
# Put your answer here
```
    git branch
    main

**Important:** You should be in the `main` branch. If you are not switch to the `main` branch

&#9989; **Question 1.3 (3 points):**
Add your name and GitHub username to the top of the notebook, then add and commit **ONLY** the notebook.

``` bash
# Put the command(s) to add and commit here 
```
git add
git commit -m 'First commit' 
git push

What is the commit message you used ?

``` bash
# Answer the question here
```
First commit

&#9989; **Question 1.4 (1 point):** Before moving on. Check that the notebook you are working on is the correct one. Run the following cell. **Are you in the new folder you just created?** If not close this notebook and open the one in the `final` folder


In [15]:
!pwd

/home/soesit/CMSE202/repositories/cmse202-f22-turnin/final


&#9989; **Question 1.5 (1 point):** Finally push the updated notebook to GitHub.

``` bash
# Put the command you used to push to GitHub here.
```
git push

**Important**: Double check you've added your Professor and your TA as collaborators to your "turnin" repository (you should have done this in the previous homework assignment).

**Also important**: Make sure that the version of this notebook that you are working on is the same one that you just added to your repository! If you are working on a different copy of the notebook, **none of your changes will be tracked**!

If everything went as intended, the file should now show up on your GitHub account in the "`cmse202-f22-turnin`" repository inside the `final` directory that you just created.


---
<a id="part2"></a>
[Back to ToC](#toc)

# Part 2. Data preprocessing (19 points)

For this assignment we’re going to be working with a dataset that contains measurements of the characteristics of multiple wines. This includes things like the acidity and density. You’ll be asked to use this information, along with the machine learning tools that we’ve used in class, to determine whether the wine type is Red or White.

The dataset is located at:
`https://raw.githubusercontent.com/msu-cmse-courses/cmse202-F22-data/main/data/winequality.csv`


**&#9989; Question 2.1 (4 points):** Do this

1. Download the data and type the command you used to download the data

2. Read the `winequality.csv` file into a dataframe

3. Print out the **unique** labels in the `type` column. 

4. Display the first 10 rows of the dataset.

**Note**: each row represents one data point and each column (except the `type` column) represents one feature. The `type` column corresponds to the class labels for every data point. There are two types of unique class labels in the `type` column. 

In [16]:
# Put the command you used to download the data here
!curl -O https://raw.githubusercontent.com/msu-cmse-courses/cmse202-F22-data/main/data/winequality.csv

  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  381k  100  381k    0     0  77332      0  0:00:05  0:00:05 --:--:-- 85552


In [17]:
# Put your code here
wine_data = pd.read_csv('winequality.csv')
wine_data.head(10)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0  white            7.0              0.27         0.36            20.7   
1  white            6.3              0.30         0.34             1.6   
2  white            8.1              0.28         0.40             6.9   
3  white            7.2              0.23         0.32             8.5   
4  white            7.2              0.23         0.32             8.5   
5  white            8.1              0.28         0.40             6.9   
6  white            6.2              0.32         0.16             7.0   
7  white            7.0              0.27         0.36            20.7   
8  white            6.3              0.30         0.34             1.6   
9  white            8.1              0.22         0.43             1.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   
5      0.050                 30.0                  97.0   0.9951  3.26   
6      0.045                 30.0                 136.0   0.9949  3.18   
7      0.045                 45.0                 170.0   1.0010  3.00   
8      0.049                 14.0                 132.0   0.9940  3.30   
9      0.044                 28.0                 129.0   0.9938  3.22   

   sulphates  alcohol  quality  
0       0.45      8.8        6  
1       0.49      9.5        6  
2       0.44     10.1        6  
3       0.40      9.9        6  
4       0.40      9.9        6  
5       0.44     10.1        6  
6       0.47      9.6        6  
7       0.45      8.8        6  
8       0.49      9.5        6  
9       0.45     11.0        6

**&#9989; Question 2.2 (7 points):** Do the following:

1. Drop the `NaN` in the dataset

2. Drop the `quality` column

2. Replace all of the strings in your `type` column with integers based on the following:

    | original type | integer type |
    | -------- | -------- |
    | white | 0 |
    | red | 1 |

3. Once you've replaced the labels, display your DataFrame and confirm that it looks correct.

In [18]:
# Put your code here
wine_data = wine_data.dropna()
wine_data = wine_data.drop(['quality'], axis=1)
wine_data['type'] = wine_data['type'].replace({'white': 0, 'red': 1})
wine_data.head(10)

type  fixed acidity  volatile acidity  citric acid  residual sugar  \
0     0            7.0              0.27         0.36            20.7   
1     0            6.3              0.30         0.34             1.6   
2     0            8.1              0.28         0.40             6.9   
3     0            7.2              0.23         0.32             8.5   
4     0            7.2              0.23         0.32             8.5   
5     0            8.1              0.28         0.40             6.9   
6     0            6.2              0.32         0.16             7.0   
7     0            7.0              0.27         0.36            20.7   
8     0            6.3              0.30         0.34             1.6   
9     0            8.1              0.22         0.43             1.5   

   chlorides  free sulfur dioxide  total sulfur dioxide  density    pH  \
0      0.045                 45.0                 170.0   1.0010  3.00   
1      0.049                 14.0                 132.0   0.9940  3.30   
2      0.050                 30.0                  97.0   0.9951  3.26   
3      0.058                 47.0                 186.0   0.9956  3.19   
4      0.058                 47.0                 186.0   0.9956  3.19   
5      0.050                 30.0                  97.0   0.9951  3.26   
6      0.045                 30.0                 136.0   0.9949  3.18   
7      0.045                 45.0                 170.0   1.0010  3.00   
8      0.049                 14.0                 132.0   0.9940  3.30   
9      0.044                 28.0                 129.0   0.9938  3.22   

   sulphates  alcohol  
0       0.45      8.8  
1       0.49      9.5  
2       0.44     10.1  
3       0.40      9.9  
4       0.40      9.9  
5       0.44     10.1  
6       0.47      9.6  
7       0.45      8.8  
8       0.49      9.5  
9       0.45     11.0

**&#9989; Question 2.3 (2 points):** As we've seen, when working with `scikit-learn` it can be much easier to work with the data if we have separate variables: one that stores the feature matrix and one that stores the class labels.

**Do This:** Split your DataFrame so that you have two separate DataFrames: (1) one called `features`, which contains all columns of features; and (2) one called `labels`, which is a single-column dataframe that contains all of the *new* integer labels you just created. 

In [41]:
# Put your code here
features = wine_data.iloc[:,1:]
labels = wine_data.iloc[:,0]
features


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0               7.0             0.270         0.36            20.7      0.045   
1               6.3             0.300         0.34             1.6      0.049   
2               8.1             0.280         0.40             6.9      0.050   
3               7.2             0.230         0.32             8.5      0.058   
4               7.2             0.230         0.32             8.5      0.058   
...             ...               ...          ...             ...        ...   
6491            6.8             0.620         0.08             1.9      0.068   
6492            6.2             0.600         0.08             2.0      0.090   
6494            6.3             0.510         0.13             2.3      0.076   
6495            5.9             0.645         0.12             2.0      0.075   
6496            6.0             0.310         0.47             3.6      0.067   

      free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                    45.0                 170.0  1.00100  3.00       0.45   
1                    14.0                 132.0  0.99400  3.30       0.49   
2                    30.0                  97.0  0.99510  3.26       0.44   
3                    47.0                 186.0  0.99560  3.19       0.40   
4                    47.0                 186.0  0.99560  3.19       0.40   
...                   ...                   ...      ...   ...        ...   
6491                 28.0                  38.0  0.99651  3.42       0.82   
6492                 32.0                  44.0  0.99490  3.45       0.58   
6494                 29.0                  40.0  0.99574  3.42       0.75   
6495                 32.0                  44.0  0.99547  3.57       0.71   
6496                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  
0         8.8  
1         9.5  
2        10.1  
3         9.9  
4         9.9  
...       ...  
6491      9.5  
6492     10.5  
6494     11.0  
6495     10.2  
6496     11.0  

[6463 rows x 11 columns]

**&#9989; Question 2.4 (1 points):** Do this:

We need to scale the data. This isn't something we've talked very much about in-class, so we don't expect you to know exactly what is happening here. But it will be important for our PCA later on in the assignment. We decided to put it here so you were aware of it's existence. 

**Run the following bit of code below.**

In [34]:
features = pd.DataFrame(PowerTransformer().fit_transform(features), 
                            columns=features.columns, 
                            index=features.index)

**&#9989; Question 2.5 (3 points):** Do this:

1. Split your data into a training and a testing set with a training set representing 75% of your data. For reproducibility , set the `random_state` argument to `314159`. 

2. Print the lengths to show you have the right number of entries for the training and testing sets.

In [21]:
X_train, X_test, y_train, y_test = train_test_split(features, labels, 
                                                    test_size=0.25, random_state=314159)
print(len(X_train))
print(len(X_test))

4847
1616


### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 2", and push the changes to GitHub.

---

<a id="part3"></a>
[Back to ToC](#toc)

# Part 3. Regression (15 Points)

**&#9989; Question 3.1.1 (3 points):**
Let's start making some prediction. 

1. Make a Logistic Regression model using `statsmodels` for predicting the type of wine. 

2. Print out the summary of your model fits.


In [24]:
# Write your code here


# Fit the model
model1 = sm.Logit(y_train, X_train)
result1 = model1.fit()

# Print the summary results
print(result1.summary())

# Predict the labels on the test data
y_pred1 = result1.predict(X_test)

         Current function value: 0.193646
         Iterations: 35
                           Logit Regression Results                           
Dep. Variable:                   type   No. Observations:                 4847
Model:                          Logit   Df Residuals:                     4837
Method:                           MLE   Df Model:                            9
Date:                Tue, 13 Dec 2022   Pseudo R-squ.:                  0.6530
Time:                        21:47:35   Log-Likelihood:                -938.60
converged:                      False   LL-Null:                       -2704.8
Covariance Type:            nonrobust   LLR p-value:                     0.000
                           coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------------
fixed acidity            1.4959      0.115     12.993      0.000       1.270       1.722
volatile acidity         1.3956    

/home/soesit/.local/lib/python3.8/site-packages/statsmodels/base/model.py:604: ConvergenceWarning: Maximum Likelihood optimization failed to converge. Check mle_retvals
  warnings.warn("Maximum Likelihood optimization failed to "


**&#9989; Question 3.1.2 (2 points):** Further examine the results by  

1. printing a classification report
2. making a confusion matrix

In [25]:

y_pred_binary1 = (y_pred > 0.5).astype(int) 
report = classification_report(y_test, y_pred_binary)
print(report)

              precision    recall  f1-score   support

           0       1.00      0.89      0.94      1216
           1       0.76      1.00      0.86       400

    accuracy                           0.92      1616
   macro avg       0.88      0.95      0.90      1616
weighted avg       0.94      0.92      0.92      1616



In [27]:
cm1 = confusion_matrix(y_test, y_pred_binary)

# Print the confusion matrix
print(cm1)

[[1088  128]
 [   1  399]]


**&#9989; Question 3.2.1 (3 points):** Make an reduced Logistic Regression model for predicting the type of wine that uses just three parameters (two from the dataset + a constant). Print out the summary of your model fits.



In [43]:

features1 = wine_data.loc[:, ['residual sugar', 'alcohol']]
X_train, X_test, y_train, y_test = train_test_split(features1, labels, 
                                                    test_size=0.25, random_state=314159)
# Add a constant term to the independent variable
X_train2 = sm.add_constant(X_train)
X_test2 = sm.add_constant(X_test)

# Make a logistic regression model
model2 = sm.Logit(y_train, X_train2)
result2 = model2.fit()


# Print the summary results
print(result2.summary())

# Predict the labels on the test data
y_pred2 = result2.predict(X_test2)

Optimization terminated successfully.
         Current function value: 0.456145
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:                   type   No. Observations:                 4847
Model:                          Logit   Df Residuals:                     4844
Method:                           MLE   Df Model:                            2
Date:                Wed, 14 Dec 2022   Pseudo R-squ.:                  0.1826
Time:                        00:00:48   Log-Likelihood:                -2210.9
converged:                       True   LL-Null:                       -2704.8
Covariance Type:            nonrobust   LLR p-value:                3.147e-215
                     coef    std err          z      P>|z|      [0.025      0.975]
----------------------------------------------------------------------------------
const              4.6351      0.376     12.319      0.000       3.898       5.373
residual sugar   

**&#9989; Question 3.2.2 (2 points):** Same as above, examine the results of your reduced model by printing a classification report and a confusion matrix

In [45]:
#Write your code here
# Convert into binary
y_pred_binary2 = (y_pred > 0.5).astype(int) 
report2 = classification_report(y_test, y_pred_binary2)
print(report2)

cm2 = confusion_matrix(y_test, y_pred_binary2)

# Print the confusion matrix
print(cm2)

              precision    recall  f1-score   support

           0       1.00      0.89      0.94      1216
           1       0.76      1.00      0.86       400

    accuracy                           0.92      1616
   macro avg       0.88      0.95      0.90      1616
weighted avg       0.94      0.92      0.92      1616

[[1088  128]
 [   1  399]]


**&#9989; Question 3.3 (3 points):** How did you pick the best parameters?

I determined mainly on the P-Value and which had the lowest P-Values. After that I basically plugged and checked. 

### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 3", and push the changes to GitHub.


---
<a id="part4"></a>
[Back to ToC](#toc)

# Part 4. Principal Component Analysis (12 points)

The full model uses all 12 features to predict the results. In many cases, we might need to see how close we can get with fewer features. Instead of simply removing features, we will use a Principal Component Analysis (PCA) to determine the combined features that contribute the most to the model (through their accounted variance).


**&#9989; Question 4.1 (5 points):** Run a Principle Component Analysis (PCA)

Since we only have 12 features to start with, let's see how well we can do if we try to aggressively reduce the feature count and use only **2** principle components. 

1. Using `PCA()` and the associated methods, run a principle component analysis on your `features` dataset using only 2 components. 

2. Transform the dataset into a new dataset and call it `pca_features`. 

3. Print the `explained_variance_ratio_` and its sum.

In [35]:
# Standardize data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
data_std = scaler.fit_transform(features)

# Fit PCA to standardized data
pca = PCA(n_components=2)
pca.fit(data_std)

# Transform data using fit_transform() method
pca_features = pca.fit_transform(data_std)

# View explained variance ratios
print(pca.explained_variance_ratio_)

# Print sum of explained variance ratios
print(pca.explained_variance_ratio_.sum())

[0.2745369 0.2286033]
0.5031401988795946


&#9989; **Question 4.1.1 (2 points):** What is the **total** explained variance ratio captured by the 2 principle components? (just quote the number) How well do you think a model with these many features will perform? Why?

<font size=+3>&#9998;</font> The total explained variance ratio captured by the 2 principle components is 0.5031401988795946.

A model with these many features may not perform reasonably well since the 2 principal components capture only half the  amount of the data's variance (50.31%). However, further evaluation of the model's performance using metrics such as accuracy, precision, and recall may be necessary to determine its actual performance.

**&#9989; Question 4.2 (3 points):** Do the following:

1. Split your new features (`pca_features`) and corresponding labels (the labels are the same as before) into a training and a testing set, with the training set representing 75% of your data. For reproducibility, set the `random_state` argument to `314159`. 

2. Print the lengths to show you have the right number of entries.

Basically do the same you have done in Q2.5 but with the `pca_features` dataset now.

In [47]:
# Put your code here
X_train, X_test, y_train, y_test = train_test_split(pca_features, labels, 
                                                    test_size=0.25, random_state=314159)
print(len(X_train))
print(len(X_test))

4847
1616


---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 4".

---

---
<a id="part5"></a>
[Back to ToC](#toc)

# Part 5. Support Vector Machine (15 points)

Let's see how an SVM performs

**&#9989; Question 5.1 (9 points):** Do the following:

1. Build a `SVC` model with a `linear` kernel. (2 points), 

2. Use `GridSearchCV` to find the best `C` parameter from this list: `'C':[0.0001, 0.001, 0.1, 1, 10]`, (2 points)
  
3. **Fit** the above model on the training set and **print** the best estimator (2 points)

4. Use your best estimator on the test dataset to make prediction. (1 point)

5. Print a [classification report](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.classification_report.html), based on its performance on the testing dataset. (1 point)

6. Print the [confusion matrix](https://scikit-learn.org/stable/auto_examples/model_selection/plot_confusion_matrix.html#sphx-glr-auto-examples-model-selection-plot-confusion-matrix-py) on the testing dataset  (1 point)

**Note:** You can set the `GridSearchCV` keyword argument `n_jobs = -1` to speed things up (this allows the process to run on multiple cores.)

**Note:** If `GridSearchCV` is slow you can use the built-in model `LinearSVC` in `sklearn`. This is the same as `SVC(kernel = 'linear')` but it is faster. 

In [49]:
# Create the SVC model with a linear kernel
model = SVC(kernel='linear')


In [51]:
start = time.time()

#make some temporary variables so you can change this easily
X_train = X_train
y_train = y_train.astype(int)

print("Fitting the classifier to the training set")
# a dictionary of hyperparameters: key is the name of the parameter, value is a list of values to test
#small gamma values
param_grid = {'C': [0.0001, 0.001, 0.1, 1, 10],
              'kernel': ['linear']}
# make a classifier by searching over a classifier and the parameter grid
clf = GridSearchCV(model, param_grid)

# we have a "good" classifier (according to GridSearchCV), how's it look
clf = clf.fit(X_train, y_train)
print("Best estimator found by grid search:")
print(clf.best_estimator_)
print("Best parameters found by grid search:")
print(clf.best_params_)

end = time.time()
print("Runtime",end - start)

Fitting the classifier to the training set
Best estimator found by grid search:
SVC(C=0.001, kernel='linear')
Best parameters found by grid search:
{'C': 0.001, 'kernel': 'linear'}
Runtime 1.6939733028411865


In [53]:
model = SVC(kernel='linear', C=0.001)

# Train the model on the training data
model.fit(X_train, y_train)

# Make predictions on the test data
y_pred3 = model.predict(X_test)


[0 0 0 ... 0 1 0]


In [54]:

y_pred_binary3 = (y_pred > 0.5).astype(int) 
report3 = classification_report(y_test, y_pred_binary3)
print(report3)

              precision    recall  f1-score   support

           0       0.99      0.99      0.99      1216
           1       0.98      0.96      0.97       400

    accuracy                           0.99      1616
   macro avg       0.98      0.98      0.98      1616
weighted avg       0.99      0.99      0.99      1616



In [55]:
cm3 = confusion_matrix(y_test, y_pred_binary3)

# Print the confusion matrix
print(cm3)

[[1208    8]
 [  16  384]]


**&#9989; Question 5.1.1 (4 points):** Answer the following questions:

1. What is the accuracy score of your model ? **Do not use code** You should be able to read it from the output above!

Accuracy socre is 0.99

2. How does your SVM model compare with the Regression models? 
 
 The percision and recall were both much more accurate

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 5".

---

<a id="part3"></a>
[Back to ToC](#toc)

# Part 6. Regression Redux (7 Points)

**&#9989; Question 6.1.1 (3 points):** Let's investigate the Logistic model again and see how it performs on the new PCA-transformed dataset. 

**Do this**: Make an Logistic model using the PCA features (2 features + constant). You should add a constant to the two PCA features, giving you three total features.


In [56]:
X_train3 = sm.add_constant(X_train)
X_test3 = sm.add_constant(X_test)

# Make a logistic regression model
model4 = sm.Logit(y_train, X_train3)
result4 = model4.fit()


# Print the summary results
print(result4.summary())

# Predict the labels on the test data
y_pred4 = result4.predict(X_test3)

Optimization terminated successfully.
         Current function value: 0.057610
         Iterations 10
                           Logit Regression Results                           
Dep. Variable:                   type   No. Observations:                 4847
Model:                          Logit   Df Residuals:                     4844
Method:                           MLE   Df Model:                            2
Date:                Wed, 14 Dec 2022   Pseudo R-squ.:                  0.8968
Time:                        00:36:52   Log-Likelihood:                -279.23
converged:                       True   LL-Null:                       -2704.8
Covariance Type:            nonrobust   LLR p-value:                     0.000
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -4.7402      0.242    -19.576      0.000      -5.215      -4.266
x1            -3.7158      0

**&#9989; Question 6.1.2 (2 points):**  Examine the results by making a confusion matrix (2 Points)

In [ ]:

y_pred_binary3 = (y_pred > 0.5).astype(int) 


cm3 = confusion_matrix(y_test, y_pred_binary3)

# Print the confusion matrix
print(cm3)

---
### &#128721; STOP
**Pause to commit your changes to your Git repository! (2 points)**

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 6".

---

---
<a id="part7"></a>
[Back to ToC](#toc)

# Part 7. Conceptual Questions (14 Points)

**&#9989; Question 7.1 (4 points):** Compare your results from 3.2.2 and 6.1.2, two Logistic regression models that used the three most significant features. Which one performed better? Why? Your explanation should discuss the difference between the features in the two models and how that difference might affect how well the model fits the data.  

*Write your response here*

#### 7.2 
The following questions probe your understanding of  `recall` and `precision`. You’ll be given a specific scenario, and you will need to decide whether to select a Machine Learning model that maximizes `recall` or `precision`.


**&#9989; Question 7.2.1 (4 points):**  A new disease has been detected in Sweden. Doctors have taken multiple measurements of patients and passed them along to you. You’ve tested various models to predict which patients have the new disease. The Swedish health experts have told you it is critical that they identify **all** of the patients with the new disease so they can put them into quarantine. They tell you that it doesn’t matter if your model accidentally flags some people as having the disease even when they don’t, as it’s much better to tell people who aren’t sick to quarantine than to tell a sick person they don’t need to quarantine.

Given this situation, should you choose the model that maximizes `recall` or the model that maximizes `precision`? Why?

In this situation, you should choose the model that maximizes recall. Recall is a measure of the ability of a model to correctly identify all patients with the disease, so it is more important to identify all patients with the disease in order to quarantine them and prevent the spread of the disease. It is better to accidentally flag some people as having the disease who don't actually have it, and tell them to quarantine, than to miss identifying patients with the disease and allowing them to spread the disease to others. Precision, on the other hand, is a measure of the ability of a model to correctly identify only those patients who actually have the disease, so it is not as important in this situation.

**&#9989; Question 7.2.2 (4 points):**
You are working for the state of Michigan, helping them with their new unemployment insurance program. You’ve created various models that are meant to detect fraud–i.e., determine whether or not someone applied for unemployment insurance when they shouldn’t have. Your boss tells you that it’s critical that your model doesn’t accidentally accuse an innocent person of fraud. They say it’s fine if your model misses some people who committed fraud as long as there are no false accusations. 

Given this situation, should you choose the model that maximizes `recall` or the model that maximizes `precision`? Why?

In this situation, you should choose the model that maximizes precision. Precision is a measure of the ability of a model to correctly identify only those people who actually committed fraud, so it is important in this situation to avoid falsely accusing innocent people. It is better to miss identifying some people who committed fraud than to falsely accuse innocent people, who could face negative consequences as a result. Recall, on the other hand, is a measure of the ability of a model to correctly identify all people who committed fraud, so it is not as important in this situation.

---
### &#128721; STOP
**Pause to commit your changes to your Git repository!** (2 points)

Take a moment to save your notebook, commit the changes to your Git repository using the commit message "Committing Part 6".

---

---
<a id="conclusion"></a>
[Back to ToC](#toc)

# Part 8. Conclusion (3 points)

Make sure all of your changes to your repository are committed and pushed to GitHub. 
Before you leave

1. Have you added your name and github username at the top of this notebook? ? (1 point)

2. Push the changes to your GitHub repository (1 point)

3. Upload your notebook to D2L in case something went wrong with your repository or if you couldn't get the repository to work.  (1 point)

## You're done! Congrats on finishing your CMSE 202 Midterm!

In [ ]:

# FINAL GRADE (Instructor's use only)",
parts = [part_0, part_1, part_2, part_3, part_4, part_5, part_6, part_7]
print(f"Part 0: {part_0}/{grades[0]}")
print(f"Part 1: {part_1}/{grades[1]}")
print(f"Part 2: {part_2}/{grades[2]}")
print(f"Part 3: {part_3}/{grades[3]}")
print(f"Part 4: {part_4}/{grades[4]}")
print(f"Part 5: {part_5}/{grades[5]}")
print(f"Part 6: {part_6}/{grades[6]}")
print(f"Part 7: {part_7}/{grades[7]}")

total = sum(parts)
print(f"Your final grade is {total}")

&#169; Copyright 2022,  Department of Computational Mathematics, Science and Engineering at Michigan State University